partiql-lang/src/main/kotlin/org/partiql/lang/syntax/impl/PartiQLPigVisitor.kt

```diff
@@ -1662,15 +1833,97 @@ internal class PartiQLPigVisitor(val customTypes: List<CustomType> = listOf(), p
+            else -> integerNode.text.toInteger().toLong()
+        }
+        if (precision < 0) {
+            throw integerNode.err("Precision out of bounds", ErrorCode.PARSE_INVALID_PRECISION_FOR_TIMESTAMP)
+        }
+        return timestampString to precision
+    }
+
+    /**
+     * Parse Timestamp based on the existence of Time zone
+     */
+    private fun getTimestampDynamic(
+        timestampString: String,
+        precision: Long?,
+        node: TerminalNode
+    ) = PartiqlAst.build {
+        val timestamp =
+            try {
+                DateTimeUtils.parseTimestamp(timestampString)
+            } catch (e: DateTimeException) {
+                throw node.err("Invalid Date Time Literal", ErrorCode.PARSE_INVALID_DATETIME_STRING, cause = e)
+            }
+        val (tzSign, tzHour, tzMinute) = getTimeZoneField(timestamp.timeZone)
+        timestamp(
+            timestampValue(
+                timestamp.year.toLong(), timestamp.month.toLong(), timestamp.day.toLong(),
+                timestamp.hour.toLong(), timestamp.minute.toLong(), ionDecimal(Decimal.valueOf(timestamp.second)),
+                tzSign, tzHour, tzMinute, precision
+            )
+        )
+    }
+
+    private fun getTimestampWithTimezone(
+        timestampString: String,
+        precision: Long?,
+        node: TerminalNode
+    ) = PartiqlAst.build {
+        val timestamp = try {
+            DateTimeUtils.parseTimestamp(timestampString)
+        } catch (e: DateTimeException) {
+            throw node.err("Invalid Date Time Literal", ErrorCode.PARSE_INVALID_DATETIME_STRING, cause = e)
+        }
+        if (timestamp.timeZone == null)
+            throw node.err(
+                "Invalid Date Time Literal, expect Time Zone for Type Timestamp With Time Zone",
+                ErrorCode.PARSE_INVALID_DATETIME_STRING
+            )
+        val (tzSign, tzHour, tzMinute) = getTimeZoneField(timestamp.timeZone)
+        timestamp(
+            timestampValue(
+                timestamp.year.toLong(), timestamp.month.toLong(), timestamp.day.toLong(),
+                timestamp.hour.toLong(), timestamp.minute.toLong(), ionDecimal(Decimal.valueOf(timestamp.second)),
+                tzSign, tzHour, tzMinute, precision
+            )
+        )
+    }
+
+    private fun getTimeZoneField(timeZone: TimeZone?) =
+        when (timeZone) {
+            TimeZone.UnknownTimeZone -> Triple("-", 0L, 0L)
+            is TimeZone.UtcOffset -> {
+                val positiveOffsetMinutes = abs(timeZone.totalOffsetMinutes)
+                val tzHour = positiveOffsetMinutes / 60
+                val tzMinutes = positiveOffsetMinutes - tzHour * 60
+                if (timeZone.totalOffsetMinutes >= 0) {
+                    Triple("+", tzHour.toLong(), tzMinutes.toLong())
+                } else {
+                    Triple("-", tzHour.toLong(), tzMinutes.toLong())
+                }
+            }
+
+            null -> Triple(null, null, null)
```

Comment:
Can you give information regarding the difference between unknown offset and null offset?